src/neo4j_genai/retrievers/text2cypher.py

```diff
@@ -0,0 +1,116 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Optional
+
+import neo4j
+from neo4j.exceptions import CypherSyntaxError, DriverError, Neo4jError
+from pydantic import ValidationError
+
+from neo4j_genai.llm import LLM
+from neo4j_genai.prompts import TEXT2CYPHER_PROMPT
+from neo4j_genai.retrievers.base import Retriever
+from neo4j_genai.schema import get_schema
+from neo4j_genai.types import (
+    LLMModel,
+    Neo4jDriverModel,
+    Neo4jSchemaModel,
+    Text2CypherRetrieverModel,
+    Text2CypherSearchModel,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Text2CypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using natural language.
+    Converts a user's natural language query to a Cypher query using an LLM,
+    then retrieves records from a Neo4j database using the generated Cypher query
+    """
+
+    def __init__(
+        self, driver: neo4j.Driver, llm: LLM, neo4j_schema: Optional[str] = None
+    ) -> None:
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            llm_model = LLMModel(llm=llm)
+            neo4j_schema_model = (
+                Neo4jSchemaModel(neo4j_schema=neo4j_schema) if neo4j_schema else None
+            )
+            validated_data = Text2CypherRetrieverModel(
+                driver_model=driver_model,
+                llm_model=llm_model,
+                neo4j_schema_model=neo4j_schema_model,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(validated_data.driver_model.driver)
+        self.llm = validated_data.llm_model.llm
+        try:
+            self.neo4j_schema = (
+                validated_data.neo4j_schema_model.neo4j_schema
+                if validated_data.neo4j_schema_model
+                else get_schema(validated_data.driver_model.driver)
+            )
+        except (Neo4jError, DriverError) as e:
```

Comment:
I think `DriverError` has no `message` attribute. To me we can just print the exception itself. Or we can use [`logger.exception(e)`](https://docs.python.org/3/library/logging.html#logging.exception)

```python
try:
    raise DriverError()
except Exception as e:
    print(e.message)
```

==>
```
Traceback (most recent call last):
  File "<ipython-input-5-2a34558e7899>", line 2, in <module>
    raise DriverError()
neo4j.exceptions.DriverError
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/estelle/.cache/pypoetry/virtualenvs/rag-evaluation-njKKJ5sT-py3.12/lib/python3.12/site-packages/IPython/core/interactiveshell.py", line 3577, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-2a34558e7899>", line 4, in <module>
    print(e.message)
          ^^^^^^^^^
AttributeError: 'DriverError' object has no attribute 'message'
```

Reply:
Updated!